src/transformation/visitors/jsx/jsx.ts

```diff
@@ -0,0 +1,252 @@
+import * as ts from "typescript";
+import { JsxEmit } from "typescript";
+import * as lua from "../../../LuaAST";
+import { FunctionVisitor, TransformationContext, Visitors } from "../../context";
+import { transformJsxAttributes } from "../literal";
+import { unsupportedJsxConfiguration } from "../../utils/diagnostics";
+import { XHTMLEntities } from "./xhtml";
+import { AnnotationKind, getFileAnnotations } from "../../utils/annotations";
+
+function checkValidJsxConfig(node: ts.Node, context: TransformationContext) {
+    if (context.options.jsx !== JsxEmit.React) {
+        context.diagnostics.push(unsupportedJsxConfiguration(node));
+    }
+}
+
+// used for jsxFactory and jsxFragmentFactory options
+function getExpressionFromOption(
+    node: ts.Node,
+    fileAnnotation: AnnotationKind,
+    option: string | undefined,
+    defaultValue: string
+): lua.Expression {
+    const annotation = getFileAnnotations(node.getSourceFile()).get(fileAnnotation);
+    const [first, second] = (annotation?.args[0] ?? option ?? defaultValue).split(".");
+    return second === undefined
+        ? lua.createIdentifier(first)
+        : lua.createTableIndexExpression(lua.createIdentifier(first), lua.createStringLiteral(second));
+}
+
+/* Implementations of jsx text processing modified from sucrase (https://github.com/alangpierce/sucrase). */
+
+const HEX_NUMBER = /^[\da-fA-F]+$/;
+const DECIMAL_NUMBER = /^\d+$/;
+
+/**
+ * Turn the given jsxText string into a JS string literal. Leading and trailing
+ * whitespace on lines is removed, except immediately after the open-tag and
+ * before the close-tag. Empty lines are completely removed, and spaces are
+ * added between lines after that.
+ *
+ * We use JSON.stringify to introduce escape characters as necessary, and trim
+ * the start and end of each line and remove blank lines.
+ */
+function formatJSXTextLiteral(text: string): string {
```

Comment:
I really do not like the prospect of having to maintain this code, does TS not already do this?

Reply:
TS does do this, but it's buried under several layers of internal code.

I adapted this implementation from sucrase instead (a babel-like processor), as it has a more straightforward implementation.